File: app/src/main/java/it/cammino/risuscito/items/InsertItem.java
package it.cammino.risuscito.items;

import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.support.annotation.ColorRes;
import android.support.annotation.StringRes;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.materialize.holder.ColorHolder;
import com.mikepenz.materialize.holder.StringHolder;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import it.cammino.risuscito.LUtils;
import it.cammino.risuscito.R;

public class InsertItem extends AbstractItem<InsertItem, InsertItem.ViewHolder> {

    private StringHolder title;
    private StringHolder page;
    private StringHolder source;
    private ColorHolder color;
    private int numSalmo;
    private String normalizedTitle;
    private String filter;
    private int id;

    public InsertItem withTitle(String title) {
        this.title = new StringHolder(title);
        return this;
    }

    public InsertItem withTitle(@StringRes int titleRes) {
        this.title = new StringHolder(titleRes);
        return this;
    }

    public InsertItem withPage(String page) {
        this.page = new StringHolder(page);
        return this;
    }

    public InsertItem withPage(@StringRes int pageRes) {
        this.page = new StringHolder(pageRes);
        return this;
    }

    public InsertItem withSource(String src) {
        this.source = new StringHolder(src);
        return this;
    }

    public InsertItem withSource(@StringRes int srcRes) {
        this.source = new StringHolder(srcRes);
        return this;
    }

    public InsertItem withColor(String color) {
        this.color = ColorHolder.fromColor(Color.parseColor(color));
        return this;
    }

    public InsertItem withColor(@ColorRes int colorRes) {
        this.color = ColorHolder.fromColorRes(colorRes);
        return this;
    }

    public InsertItem withId(int id) {
        this.id = id;
        super.withIdentifier(id);
        return this;
    }

    public InsertItem withNumSalmo(String numSalmo) {
        int numeroTemp = 0;
        try {
            numeroTemp = Integer.valueOf(numSalmo.substring(0, 3));
        } catch (NumberFormatException | IndexOutOfBoundsException e) {
            Log.e(getClass().getName(), e.getLocalizedMessage(), e);
        }
        this.numSalmo = numeroTemp;
        return this;
    }

    public InsertItem withNormalizedTitle(String normTitle) {
        this.normalizedTitle = normTitle;
        return this;
    }

    public InsertItem withFilter(String filter) {
        this.filter = filter;
        return this;
    }

    public StringHolder getTitle() {
        return title;
    }

    public StringHolder getPage() {
        return page;
    }

    public StringHolder getSource() {
        return source;
    }

    public ColorHolder getColor() {
        return color;
    }

    public int getId() {
        return id;
    }

    public int getNumSalmo() {
        return numSalmo;
    }

    /**
     * defines the type defining this item. must be unique. preferably an id
     *
     * @return the type
     */
    @Override
    public int getType() {
        return R.id.fastadapter_insert_item_id;
    }

    /**
     * defines the layout which will be used for this item in the list
     *
     * @return the layout for this item
     */
    @Override
    public int getLayoutRes() {
        return R.layout.row_item_to_insert;
    }

    /**
     * binds the data of this item onto the viewHolder
     *
     * @param viewHolder the viewHolder of this item
     */
    @Override
    public void bindView(ViewHolder viewHolder, List<Object> payloads) {
        super.bindView(viewHolder, payloads);

        //set the text for the name
        if (filter != null && !filter.isEmpty()) {
            int mPosition = normalizedTitle.indexOf(filter);
            if (mPosition >= 0) {
                String highlighted = title.getText().replaceAll("(?i)(" + title.getText().substring(mPosition, mPosition + filter.length()) + ")", "<b>$1</b>");
                viewHolder.mTitle.setText(LUtils.fromHtmlWrapper(highlighted));
            }
            else
                StringHolder.applyTo(title, viewHolder.mTitle);
        }
        else
            StringHolder.applyTo(title, viewHolder.mTitle);
        //set the text for the description or hide
        StringHolder.applyToOrHide(page, viewHolder.mPage);
        GradientDrawable bgShape = (GradientDrawable) viewHolder.mPage.getBackground();
        bgShape.setColor(color.getColorInt());

    }

    @Override
    public void unbindView(ViewHolder holder) {
        super.unbindView(holder);
        holder.mTitle.setText(null);
        holder.mPage.setText(null);
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    /**
     * our ViewHolder
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        protected View view;
        @BindView(R.id.text_title) TextView mTitle;
        @BindView(R.id.text_page) TextView mPage;
        public @BindView(R.id.preview) View mPreview;

        public ViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            this.view = view;
        }
    }

}

File: app/src/main/java/it/cammino/risuscito/items/InsertItem.kt
package it.cammino.risuscito.items

import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.ViewGroup
import com.mikepenz.fastadapter.binding.AbstractBindingItem
import com.mikepenz.fastadapter.ui.utils.StringHolder
import it.cammino.risuscito.R
import it.cammino.risuscito.databinding.RowItemToInsertBinding
import it.cammino.risuscito.utils.StringUtils
import it.cammino.risuscito.utils.Utility
import it.cammino.risuscito.utils.Utility.helperSetColor
import it.cammino.risuscito.utils.Utility.helperSetString
import it.cammino.risuscito.utils.extension.spannedFromHtml
import it.cammino.risuscito.utils.extension.systemLocale

fun insertItem(block: InsertItem.() -> Unit): InsertItem = InsertItem().apply(block)

class InsertItem : AbstractBindingItem<RowItemToInsertBinding>() {

    var title: StringHolder? = null
        private set
    var setTitle: Any? = null
        set(value) {
            title = helperSetString(value)
        }
    var page: StringHolder? = null
        private set
    var setPage: Any? = null
        set(value) {
            page = helperSetString(value)
        }
    var source: StringHolder? = null
        private set
    var setSource: Any? = null
        set(value) {
            source = helperSetString(value)
        }
    var undecodedSource: String? = null
    var color: Int = Color.WHITE
        private set
    var setColor: String? = null
        set(value) {
            color = helperSetColor(value)
        }
    var filter: String? = null
    var id: Int = 0
        set(value) {
            identifier = value.toLong()
            field = value
        }
    var consegnato: Int = 0

    override val type: Int
        get() = R.id.fastadapter_insert_item_id

    override fun createBinding(
        inflater: LayoutInflater,
        parent: ViewGroup?
    ): RowItemToInsertBinding {
        return RowItemToInsertBinding.inflate(inflater, parent, false)
    }

    override fun bindView(binding: RowItemToInsertBinding, payloads: List<Any>) {
        // get the context
        val ctx = binding.root.context

        //set the text for the name
        filter?.let {
            if (it.isNotEmpty()) {
                val normalizedTitle = Utility.removeAccents(
                    title?.getText(ctx).orEmpty()
                )
                val mPosition =
                    normalizedTitle.lowercase(ctx.resources.systemLocale).indexOf(it)
                if (mPosition >= 0) {
                    val stringTitle = title?.getText(ctx)
                    val highlighted = StringBuilder(
                        if (mPosition > 0) (stringTitle?.substring(0, mPosition)
                            .orEmpty()) else StringUtils.EMPTY
                    )
                        .append("<b>")
                        .append(stringTitle?.substring(mPosition, mPosition + it.length))
                        .append("</b>")
                        .append(stringTitle?.substring(mPosition + it.length))
                    binding.textTitle.text = highlighted.toString().spannedFromHtml
                } else
                    StringHolder.applyTo(title, binding.textTitle)
            } else
                StringHolder.applyTo(title, binding.textTitle)
        } ?: StringHolder.applyTo(title, binding.textTitle)

        //set the text for the description or hide
        StringHolder.applyToOrHide(page, binding.textPage)
        val bgShape = binding.textPage.background as? GradientDrawable
        bgShape?.setColor(color)
    }

    override fun unbindView(binding: RowItemToInsertBinding) {
        binding.textTitle.text = null
        binding.textPage.text = null
    }

}